### PR TITLE
Cs/sc 3059 group cases manually

### PR DESCRIPTION
**Pull request by Charl1996**
## Product Description
This PR adds the ability to modify the case groupings manually on the map by hovering over the cases. See below for an example:

manual_case_grouping

In addition, the respective case group colors are displayed next to the groups in the table under the map for easier reference, as seen below.

image

## Technical Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3059)

A special case group is added for all cases not belonging to a group on the map. This makes it easier to handle the reassigning between groups.

When the user clicks the Export Groups button only cases belonging to groups will be exported, _excluding_ the special group (which represents the "No Group" lot).

Also, the groups displayed on the screen will now be displayed by generated group name (`Group 1`, `Group 2`, etc) instead of an UUID so that it's less of a mental burden for users to keep track of where each group is on the map. When the user regenerates the groups the group names will be regenerated as well which means the group names does not persist across generating groups. I think this is OK since 
1. It's pretty hard to try and persist group names, especially since group definitions might differ between generation attempts
2. It's pretty obvious which groups are which once generated

## Feature Flag
Geospatial

### Safety story
Tested locally

### Automated test coverage
No automated tests for this.

### QA Plan
QA to follow when feature is slightly more mature.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by Charl1996** [comment]
Again, seems like eslint does not like `async` function definitions.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js [hunk 1]
```diff
@@ -102,17 +110,29 @@ hqDefine("geospatial/js/case_grouping_map",[
             hiddenElement.remove();
         };
 
-        self.addGroupsToCases = function(caseGroups) {
-            clearCaseGroups();
+        self.addGroupDataToCases = function(caseGroups, groupsData, assignDefaultGroup) {
+            const defaultGroup = groupsData[0];
```
Comment (mkangia) [inline corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:114]: are we assuming there that the first group is the unassigned one?

**Review by mkangia — APPROVED** [review]

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js (hunk 1 above)
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:114]: @mkangia Correct.

Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:114]: is there a more reliable way to do that instead of using index?

Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:114]: Yeah, I suppose so. I haven't given it much attention after I wrote it, but maybe I should. I'll update.

Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js]: https://github.com/dimagi/commcare-hq/pull/33735/commits/1f79cf5f35397123ac60fcd3ff6052c354cc65c5

Comment (mkangia) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js]: Thank you.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js [hunk 2]
```diff
@@ -441,30 +535,47 @@ hqDefine("geospatial/js/case_grouping_map",[
...
             const pointCount = cluster.properties.point_count;
 
             try {
                 const casePoints = await getClusterLeavesAsync(clusterSource, clusterId, pointCount);
                 const groupUUID = uuidv4();
-                for (const casePoint of casePoints) {
-                    const caseId = casePoint.properties.id;
-                    caseGroups[caseId] = {
+
+                if (casePoints.length > 0) {
```
Comment (zandre-eng) [inline corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:560]: Nit: This can be shortened to `if (casePoints.length)`

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js [hunk 3]
```diff
@@ -320,28 +377,64 @@ hqDefine("geospatial/js/case_grouping_map",[
...
     }
 
     function caseGroupSelectModel() {
         'use strict';
         var self = {};
 
-        self.allGroups = ko.observableArray([]);
         self.allCaseGroups;
+        // allGroups and caseGroupsForTable contains the same data, but there's weird knockoutjs behaviour
+        // if we're making allGroups an observable. caseGroupsForTable is populated by setCaseGroupsForTable
```
Comment (zandre-eng) [inline corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:435]: I'm curious about this, what is the weird behaviour if `allGroups` is an `observableArray`?

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js [hunk 4]
```diff
@@ -320,28 +377,64 @@ hqDefine("geospatial/js/case_grouping_map",[
...
+        self.allGroups = [];
```
Comment (zandre-eng) [inline corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:436]: It was a little difficult to initially understand the difference between `allGroups` and `allCaseGroups` since their variable names makes them sound very similar. Could they perhaps have more distinct names for what they represent? A comment explaining the difference could also work.

File: corehq/apps/geospatial/templates/case_grouping_map.html [hunk 5]
```diff
@@ -88,12 +88,12 @@
         </div>
         <div class="row">
           <div class="col col-md-6">
-            <button class="btn-default form-control" data-bind="click: showSelectedGroups()">
+            <button class="btn-default form-control" data-bind="click: showSelectedGroups(), disable: !groupsReady()">
```
Comment (zandre-eng) [inline corehq/apps/geospatial/templates/case_grouping_map.html:91]: This looks like it might give a ko exception when executed. The `exportModel` ko model has been binded to the `export-controls` element, but this button element is currently not in scope for that element. This button wouldn't have access to `groupsReady()` from `exportModel`.

File: corehq/apps/geospatial/templates/case_grouping_map.html [hunk 6]
```diff
@@ -101,4 +102,15 @@
     </div>
   </div>
 </div>
+
+<script type="text/html" id="select-case">
+  <div class="d-flex flex-row">
+    <label data-bind="attr: {for: selectCssId}">Select group</label>
```
Comment (zandre-eng) [inline corehq/apps/geospatial/templates/case_grouping_map.html:108]: This needs a trans block.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js [hunk 7]
```diff
@@ -441,30 +535,47 @@ hqDefine("geospatial/js/case_grouping_map",[
...
-        processedCluster = {}
+        processedCluster = {};
+
+        var groupCount = 1;
+        var groups = [DEFAULT_GROUP];
 
         for (const cluster of sourceFeatures) {
             const clusterId = cluster.properties.cluster_id;
-            if (!processedCluster[clusterId]) {
+            if (processedCluster[clusterId] == undefined) {
```
Comment (zandre-eng) [inline corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:547]: Could we change this to a strict equality operator (`===`)?

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js [hunk 8]
```diff
@@ -320,28 +377,64 @@ hqDefine("geospatial/js/case_grouping_map",[
...
+
+                let popup = new mapboxgl.Popup({ offset: 25, anchor: "bottom" })  // eslint-disable-line no-undef
+                    .setLngLat(coordinates)
+                    .setDOMContent(popupDiv);
+
+                marker.setPopup(popup);
+
+                const markerDiv = marker.getElement();
+                // Show popup on hover
+                markerDiv.addEventListener('mouseenter', () => marker.togglePopup());
```
Comment (zandre-eng) [inline corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:399]: Nit: Since we aren't passing any params, we can shorten this to `markerDiv.addEventListener('mouseenter', marker.togglePopup)`

**Review by zandre-eng — COMMENTED** [review]
Looking good! Just a few comments from my end. I also had a thought after watching the demo. Should we have the coloured square in the dropdown select as well? This would make it a bit easier to see which group belongs where when manually changing the group for a case.

File: corehq/apps/geospatial/templates/case_grouping_map.html (hunk 5 above)
Comment (Charl1996) [reply corehq/apps/geospatial/templates/case_grouping_map.html:115]: `groupsReady` is defined on `exportModel` and `caseGroupSelectModel`

Comment (zandre-eng) [reply corehq/apps/geospatial/templates/case_grouping_map.html:115]: Ah! It seems I completely missed that, thanks for clarifying.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js (hunk 8 above)
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:289]: Changing it to the suggestion will result in the function executing on load instead of executing the function on mouse enter.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js (hunk 3 above)
Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js]: When `allGroups` is an observable an issue happened where groups would get updated in the background seemingly randomly when you show only a subset of groups on the screen and then select to show all groups on the screen. The groups that was obscured would then either have all been unassigned or mostly unassigned from their initial groups. 

I haven't been able to figure out exactly why, as the logic for handling the selected groups on the screen seems solid, so my conclusion was that somewhere something listens to this observable which triggers a chain reaction which is hard to follow.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js (hunk 8 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:289]: That's interesting, is it because we are passing the function from the `marker` object instead of passing a function directly (e.g. `('mouseenter', fooFunc)`)? 

Looking at [documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) online, `addEventListener` should be able to take just the function object itself as a param. The code is fine as is, but I'm just curious what would cause it to execute on load instead of when the event happens.

Comment (Charl1996) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js:289]: O, wait, I just realized I was being a dumb-dumb. My mistake. I added the brackets which ran the function instead of just passing in the function name.

File: corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/case_grouping_map.js]: Hmm... yeah this definitely smells of some funky JS behaviour. :thinking: 

Since JS handles an array as pass-by-reference, I wonder if the array was perhaps being pased somewhere else as well and, instead of copying the array, it was modifying the old reference directly.

**Review by zandre-eng — APPROVED** [review]
Thanks for the refactors! This looks good :rocket:
